File: tideform-expo/lib/indexer.ts
/**
 * lib/indexer.ts — read-side queries. NO backend required (source-of-truth §12):
 * events, objects, and Walrus blobs are all read straight from the device against
 * public endpoints.
 *
 * Contract surface:
 *   listFormsForOwner(addr), fetchForm(id), fetchFormSchema(blobId),
 *   listSubmissions(formId), fetchSubmissionPayload(blobId)
 *
 * KEY DETAILS:
 *   - Event TYPE queries use the ORIGINAL package ID (env.originalPackageId) —
 *     event type-origin never changes across upgrades.
 *   - On-chain `vector<u8>` blob-ID fields are ASCII bytes of the blob-ID string.
 *     Decode with UTF-8 (decodeAsciiBlobId). NEVER base64-decode them.
 */

import type { SuiEvent, SuiObjectResponse } from '@mysten/sui/client';

import { env } from './env';
import type { FormSchema, Submission } from './schema';
import { suiClient } from './sui';
import { readJson } from './walrus';

const FORM_CREATED = `${env.originalPackageId}::events::FormCreated`;
const SUBMISSION_RECEIVED = `${env.originalPackageId}::events::SubmissionReceived`;

const MULTIGET_CHUNK = 50; // Sui multiGetObjects cap per call.
const MAX_EVENT_PAGES = 50; // safety bound for the teaching app.

// ── Parsed on-chain object shapes ─────────────────────────────────────────────

export interface FormObject {
  id: string;
  owner: string;
  admins: string[];
  /** Decoded Walrus blob ID of the form schema JSON. */
  schemaBlobId: string;
  createdAtMs: number;
  updatedAtMs: number;
  version: number;
  /** 0 OPEN · 1 CLOSED · 2 ARCHIVED */
  status: number;
  submissionsCount: number;
  requireWallet: boolean;
  onePerWallet: boolean;
}

export interface SubmissionObject {
  id: string;
  formId: string;
  /** Decoded Walrus blob ID of the submission payload JSON. */
  blobId: string;
  submitter: string;
  submittedAtMs: number;
  /** 0 NEW · 1 IN_PROGRESS · 2 RESOLVED · 3 SPAM */
  status: number;
  /** 0 LOW · 1 MED · 2 HIGH · 3 URGENT */
  priority: number;
  tags: string[];
  hasNotes: boolean;
  /** Decoded Walrus blob ID of admin notes, or "" if none. */
  notesBlobId: string;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/**
 * Decode an on-chain ASCII `vector<u8>` blob-ID field to its string form.
 * Accepts the JSON-RPC number[] representation, a Uint8Array, or a string that
 * is already decoded (passed through unchanged). NEVER base64-decode.
 */
export function decodeAsciiBlobId(
  raw: number[] | Uint8Array | string | null | undefined,
): string {
  if (raw == null) return '';
  if (typeof raw === 'string') return raw; // already the ASCII string
  const bytes = raw instanceof Uint8Array ? raw : Uint8Array.from(raw);
  return new TextDecoder().decode(bytes);
}

function num(v: unknown): number {
  if (typeof v === 'number') return v;
  if (typeof v === 'string') return Number(v);
  return 0;
}

/** Pull the `fields` bag out of a moveObject response, or null. */
function moveFields(res: SuiObjectResponse): Record<string, any> | null {
  const content = res.data?.content;
  if (!content || content.dataType !== 'moveObject') return null;
  return (content as { fields: Record<string, any> }).fields ?? null;
}

function objectId(res: SuiObjectResponse): string {
  return res.data?.objectId ?? '';
}

function parseFormObject(res: SuiObjectResponse): FormObject | null {
  const f = moveFields(res);
  if (!f) return null;
  // VecSet<address> serializes as { fields: { contents: [...] } }.
  const admins: string[] = Array.isArray(f.admins?.fields?.contents)
    ? f.admins.fields.contents
    : Array.isArray(f.admins?.contents)
      ? f.admins.contents
      : [];
  return {
    id: objectId(res) || f.id?.id || '',
    owner: f.owner,
    admins,
    schemaBlobId: decodeAsciiBlobId(f.schema_blob_id),
    createdAtMs: num(f.created_at_ms),
    updatedAtMs: num(f.updated_at_ms),
    version: num(f.version),
    status: num(f.status),
    submissionsCount: num(f.submissions_count),
    requireWallet: Boolean(f.require_wallet),
    onePerWallet: Boolean(f.one_per_wallet),
  };
}

function parseSubmissionObject(res: SuiObjectResponse): SubmissionObject | null {
  const s = moveFields(res);
  if (!s) return null;
  return {
    id: objectId(res) || s.id?.id || '',
    formId: s.form_id,
    blobId: decodeAsciiBlobId(s.blob_id),
    submitter: s.submitter,
    submittedAtMs: num(s.submitted_at_ms),
    status: num(s.status),
    priority: num(s.priority),
    tags: Array.isArray(s.tags) ? s.tags : [],
    hasNotes: Boolean(s.has_notes),
    notesBlobId: decodeAsciiBlobId(s.notes_blob_id),
  };
}

/** Query ALL events of a type (paginated, descending), with a safety bound. */
async function queryAllEvents(moveEventType: string): Promise<SuiEvent[]> {
  const all: SuiEvent[] = [];
  let cursor: { txDigest: string; eventSeq: string } | null | undefined = null;
  for (let page = 0; page < MAX_EVENT_PAGES; page += 1) {
    const res = await suiClient.queryEvents({
      query: { MoveEventType: moveEventType },
      cursor: cursor ?? null,
      limit: 50,
      order: 'descending',
    });
    all.push(...res.data);
    if (!res.hasNextPage || !res.nextCursor) break;
    cursor = res.nextCursor;
  }
  return all;
}

async function multiGet(ids: string[]): Promise<SuiObjectResponse[]> {
  const out: SuiObjectResponse[] = [];
  for (let i = 0; i < ids.length; i += MULTIGET_CHUNK) {
    const chunk = ids.slice(i, i + MULTIGET_CHUNK);
    const res = await suiClient.multiGetObjects({
      ids: chunk,
      options: { showContent: true, showType: true },
    });
    out.push(...res);
  }
  return out;
}

// ── Contract surface ──────────────────────────────────────────────────────────

/**
 * Forms owned by `addr`: query FormCreated, keep those whose `owner` matches,
 * then multiGetObjects to read current Form state. (Flow B.)
 */
export async function listFormsForOwner(addr: string): Promise<FormObject[]> {
  const events = await queryAllEvents(FORM_CREATED);
  const ids: string[] = [];
  const seen = new Set<string>();
  for (const ev of events) {
    const pj = ev.parsedJson as { form_id?: string; owner?: string } | undefined;
    if (!pj?.form_id || pj.owner !== addr) continue;
    if (seen.has(pj.form_id)) continue;
    seen.add(pj.form_id);
    ids.push(pj.form_id);
  }
  if (ids.length === 0) return [];
  const objects = await multiGet(ids);
  return objects
    .map(parseFormObject)
    .filter((f): f is FormObject => f !== null);
}

/** Read a single Form object's current on-chain state. */
export async function fetchForm(id: string): Promise<FormObject | null> {
  const res = await suiClient.getObject({
    id,
    options: { showContent: true, showType: true },
  });
  return parseFormObject(res);
}

/** Fetch + parse the form schema JSON from Walrus by its blob ID. */
export async function fetchFormSchema(blobId: string): Promise<FormSchema> {
  return readJson<FormSchema>(blobId);
}

/**
 * Submissions for a form: query SubmissionReceived filtered by `form_id`, then
 * multiGetObjects to read current Submission state (status/priority/tags). (Flow E.)
 */
export async function listSubmissions(formId: string): Promise<SubmissionObject[]> {
  const events = await queryAllEvents(SUBMISSION_RECEIVED);
  const ids: string[] = [];
  const seen = new Set<string>();
  for (const ev of events) {
    const pj = ev.parsedJson as
      | { form_id?: string; submission_id?: string }
      | undefined;
    if (!pj?.submission_id || pj.form_id !== formId) continue;
    if (seen.has(pj.submission_id)) continue;
    seen.add(pj.submission_id);
    ids.push(pj.submission_id);
  }
  if (ids.length === 0) return [];
  const objects = await multiGet(ids);
  return objects
    .map(parseSubmissionObject)
    .filter((s): s is SubmissionObject => s !== null);
}

/** Fetch + parse a submission payload JSON from Walrus by its blob ID. */
export async function fetchSubmissionPayload(blobId: string): Promise<Submission> {
  return readJson<Submission>(blobId);
}

File: tideform-swift/Sources/Tideform/Lib/Indexer.swift
//
//  Indexer.swift
//  Tideform · Lib layer
//
//  Read-side glue, mirroring the Expo `lib/indexer.ts` surface:
//    listFormsForOwner(addr), fetchForm(id), fetchFormSchema(blobId),
//    listSubmissions(formId), fetchSubmissionPayload(blobId)
//
//  Pattern (source-of-truth §9): query origin-package events -> multiGetObjects ->
//  decode Move struct fields -> read the referenced Walrus blobs. All on-device,
//  no backend, no auth.
//
//  Blob IDs in `schema_blob_id` / `blob_id` are `vector<u8>` of ASCII bytes — decoded
//  with UTF-8 here (source-of-truth §4/§12), never base64.
//

import Foundation

// MARK: - On-chain object models (distinct from the JSON `Submission` payload in Schema.swift)

/// Decoded `tideform::form::Form` object (mirrors the Expo `FormObject`; named to
/// avoid shadowing `SwiftUI.Form` in the UI layer and to match the sibling `SubmissionObject`).
public struct FormObject: Sendable, Identifiable {
    public let id: String
    public let owner: String
    public let admins: [String]
    /// ASCII-decoded Walrus blob ID of the form's `FormSchema` JSON.
    public let schemaBlobId: String
    public let version: UInt64
    public let status: UInt8          // 0 OPEN · 1 CLOSED · 2 ARCHIVED
    public let submissionsCount: UInt64
    public let requireWallet: Bool
    public let onePerWallet: Bool
    public let createdAtMs: UInt64
    public let updatedAtMs: UInt64
}

/// Decoded `tideform::submission::Submission` object.
public struct SubmissionObject: Sendable, Identifiable {
    public let id: String
    public let formId: String
    /// ASCII-decoded Walrus blob ID of the submission's `Submission` JSON payload.
    public let blobId: String
    public let submitter: String
    public let submittedAtMs: UInt64
    public let status: UInt8          // 0 NEW · 1 IN_PROGRESS · 2 RESOLVED · 3 SPAM
    public let priority: UInt8        // 0 LOW · 1 MED · 2 HIGH · 3 URGENT
    public let tags: [String]
    public let hasNotes: Bool
    /// ASCII-decoded Walrus blob ID of the admin notes (empty when `hasNotes == false`).
    public let notesBlobId: String
}

public enum IndexerError: Error, CustomStringConvertible {
    case notFound(String)
    case malformed(String)

    public var description: String {
        switch self {
        case .notFound(let s): return "Not found: \(s)"
        case .malformed(let s): return "Malformed object: \(s)"
        }
    }
}

public final class Indexer: @unchecked Sendable {

    public static let shared = Indexer()

    private let sui: SuiClient
    private let walrus: Walrus

    public init(sui: SuiClient = .shared, walrus: Walrus = .shared) {
        self.sui = sui
        self.walrus = walrus
    }

    // Event types are queried against the ORIGINAL package id (type-origin is upgrade-stable).
    private var formCreatedType: String { "\(env.originalPackageId)::events::FormCreated" }
    private var submissionReceivedType: String {
        "\(env.originalPackageId)::events::SubmissionReceived"
    }

    // MARK: - Forms

    /// Forms owned by `owner`: FormCreated events filtered by owner -> multiGetObjects.
    public func listFormsForOwner(_ owner: String) async throws -> [FormObject] {
        let target = Indexer.normalizeAddress(owner)
        let events = try await sui.queryAllEvents(eventType: formCreatedType)

        // Preserve newest-first order while de-duping form ids.
        var seen = Set<String>()
        var formIds: [String] = []
        for ev in events {
            guard
                let pj = ev.parsedJson,
                Indexer.normalizeAddress(pj["owner"]?.stringValue ?? "") == target,
                let fid = pj["form_id"]?.stringValue
            else { continue }
            if seen.insert(fid).inserted { formIds.append(fid) }
        }
        guard !formIds.isEmpty else { return [] }

        let responses = try await multiGetChunked(formIds)
        return responses.compactMap { Indexer.parseForm($0) }
    }

    /// Fetch and decode a single Form object.
    public func fetchForm(_ id: String) async throws -> FormObject {
        let response = try await sui.getObject(id: id)
        guard let form = Indexer.parseForm(response) else {
            throw IndexerError.notFound("Form \(id)")
        }
        return form
    }

    /// Read a form's `FormSchema` JSON from Walrus by its (ASCII) blob ID.
    public func fetchFormSchema(_ blobId: String) async throws -> FormSchema {
        try await walrus.readJson(FormSchema.self, id: blobId)
    }

    // MARK: - Submissions

    /// Submissions for a form: SubmissionReceived events filtered by form_id -> multiGetObjects.
    public func listSubmissions(_ formId: String) async throws -> [SubmissionObject] {
        let target = Indexer.normalizeAddress(formId)
        let events = try await sui.queryAllEvents(eventType: submissionReceivedType)

        var seen = Set<String>()
        var ids: [String] = []
        for ev in events {
            guard
                let pj = ev.parsedJson,
                Indexer.normalizeAddress(pj["form_id"]?.stringValue ?? "") == target,
                let sid = pj["submission_id"]?.stringValue
            else { continue }
            if seen.insert(sid).inserted { ids.append(sid) }
        }
        guard !ids.isEmpty else { return [] }

        let responses = try await multiGetChunked(ids)
        return responses.compactMap { Indexer.parseSubmission($0) }
    }

    /// `sui_multiGetObjects` caps at ~50 ids per call; chunk to stay under it
    /// (mirrors the Expo `MULTIGET_CHUNK = 50`).
    private func multiGetChunked(_ ids: [String], chunk: Int = 50) async throws
        -> [SuiObjectResponse]
    {
        var out: [SuiObjectResponse] = []
        var i = 0
        while i < ids.count {
            let slice = Array(ids[i..<min(i + chunk, ids.count)])
            out.append(contentsOf: try await sui.multiGetObjects(ids: slice))
            i += chunk
        }
        return out
    }

    /// Read a submission's `Submission` JSON payload from Walrus by its (ASCII) blob ID.
    public func fetchSubmissionPayload(_ blobId: String) async throws -> Submission {
        try await walrus.readJson(Submission.self, id: blobId)
    }

    // MARK: - Move struct parsing

    static func parseForm(_ response: SuiObjectResponse) -> FormObject? {
        guard let data = response.data, let fields = data.content?.fields else { return nil }
        guard
            let owner = fields["owner"]?.stringValue,
            let schemaBlobId = decodeAsciiBlob(fields["schema_blob_id"])
        else { return nil }

        return FormObject(
            id: data.objectId,
            owner: owner,
            admins: parseAddressSet(fields["admins"]),
            schemaBlobId: schemaBlobId,
            version: fields["version"]?.uint64Value ?? 0,
            status: fields["status"]?.uint8Value ?? 0,
            submissionsCount: fields["submissions_count"]?.uint64Value ?? 0,
            requireWallet: fields["require_wallet"]?.boolValue ?? false,
            onePerWallet: fields["one_per_wallet"]?.boolValue ?? false,
            createdAtMs: fields["created_at_ms"]?.uint64Value ?? 0,
            updatedAtMs: fields["updated_at_ms"]?.uint64Value ?? 0)
    }

    static func parseSubmission(_ response: SuiObjectResponse) -> SubmissionObject? {
        guard let data = response.data, let fields = data.content?.fields else { return nil }
        guard
            let formId = fields["form_id"]?.stringValue,
            let blobId = decodeAsciiBlob(fields["blob_id"]),
            let submitter = fields["submitter"]?.stringValue
        else { return nil }

        return SubmissionObject(
            id: data.objectId,
            formId: formId,
            blobId: blobId,
            submitter: submitter,
            submittedAtMs: fields["submitted_at_ms"]?.uint64Value ?? 0,
            status: fields["status"]?.uint8Value ?? 0,
            priority: fields["priority"]?.uint8Value ?? 0,
            tags: parseStringVector(fields["tags"]),
            hasNotes: fields["has_notes"]?.boolValue ?? false,
            notesBlobId: decodeAsciiBlob(fields["notes_blob_id"]) ?? "")
    }

    /// `VecSet<address>` renders as `{ "fields": { "contents": ["0x..", ...] } }`
    /// (source-of-truth §4: `f.admins.fields.contents`). Falls back to a direct array.
    static func parseAddressSet(_ value: JSONValue?) -> [String] {
        guard let value else { return [] }
        if let contents = value["fields"]?["contents"]?.arrayValue {
            return contents.compactMap { $0.stringValue }
        }
        if let arr = value.arrayValue {
            return arr.compactMap { $0.stringValue }
        }
        return []
    }

    static func parseStringVector(_ value: JSONValue?) -> [String] {
        value?.arrayValue?.compactMap { $0.stringValue } ?? []
    }

    /// Decode a Move `vector<u8>` blob-ID field to its ASCII string (source-of-truth §4/§12).
    static func decodeAsciiBlob(_ value: JSONValue?) -> String? {
        guard let value else { return nil }
        // Primary path: fullnode renders `vector<u8>` as a JSON number array.
        if let bytes = value.byteArrayValue {
            return bytes.isEmpty ? "" : String(decoding: bytes, as: UTF8.self)
        }
        // Defensive: some setups render `vector<u8>` as a base64 string; the underlying
        // bytes are the ASCII blob ID. (Blob IDs are base64URL, which standard base64
        // decoding rejects -> we fall through and return the string unchanged.)
        if let s = value.stringValue {
            if let decoded = Data(base64Encoded: s),
                let str = String(data: decoded, encoding: .utf8), !str.isEmpty
            {
                return str
            }
            return s
        }
        return nil
    }

    /// Lowercase for case-insensitive address/ID comparison. Fullnode returns full-length
    /// 0x-padded addresses, matching the values from `/api/auth/me`.
    static func normalizeAddress(_ s: String) -> String {
        s.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }
}

/// Module-level alias mirroring the JS `indexer` singleton (`indexer.listFormsForOwner(...)`).
public let indexer = Indexer.shared
